url/src/parser.rs

```diff
@@ -1473,37 +1473,81 @@ impl<'a> Parser<'a> {
         &mut self,
         scheme_type: SchemeType,
         scheme_end: u32,
-        mut input: Input<'i>,
+        input: Input<'i>,
     ) -> Option<Input<'i>> {
-        let len = input.chars.as_str().len();
-        let mut query = String::with_capacity(len); // FIXME: use a streaming decoder instead
-        let mut remaining = None;
-        while let Some(c) = input.next() {
-            if c == '#' && self.context == Context::UrlParser {
-                remaining = Some(input);
-                break;
-            } else {
-                self.check_url_code_point(c, &input);
-                query.push(c);
+        struct QueryPartIter<'i, 'p> {
+            is_url_parser: bool,
+            input: Input<'i>,
+            violation_fn: Option<&'p dyn Fn(SyntaxViolation)>,
+        }
+
+        impl<'i> Iterator for QueryPartIter<'i, '_> {
+            type Item = (&'i str, bool);
+
+            fn next(&mut self) -> Option<Self::Item> {
+                let start = self.input.chars.as_str();
+                // bypass self.input.next() in order to get string slices
+                // which are faster to operate on
+                while let Some(c) = self.input.chars.next() {
+                    match c {
+                        '\t' | '\n' | '\r' => {
```

Comment:
Can you extract this pattern to be shared with `self.input.next()` so that it is clear if this ever gets updated?

Reply:
Makes it 10ns slower even with `#[inline]` for some reason (my guess is that causes it to lose out on some optimization in the match). It seems this pattern is used in a lot of places.

Reply:
You can do
```
macro_rules! some_name {
  () => '\t' | '\n' | '\r'
}
```

and then use `some_name!()` instead of the pattern here. That should have no runtime cost.

Reply:
Thanks! That worked.